src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala

```diff
@@ -69,19 +69,23 @@ object PV63 {
             RLPList(RLPValue(hpEncode(keyNibbles.toArray[Byte], isLeaf = false)), child.fold(_.hash, _.value): RLPEncodeable)
           case n: MptBranch =>
             import n._
-            RLPList(children.map(e => e.fold(_.hash, _.value)).map(e => byteStringEncDec.encode(e)) :+ (terminator: RLPEncodeable): _*)
+            RLPList(children.map(e => e.fold(a => byteStringEncDec.encode(a.hash), b => rawDecode(b.value.toArray[Byte]))) :+ (terminator: RLPEncodeable): _*)
         }
       }
 
       override def decode(rlp: RLPEncodeable): MptNode = rlp match {
         case rlpList: RLPList if rlpList.items.length == BranchNodeChildLength + 1 =>
-          MptBranch(rlpList.items.take(BranchNodeChildLength).map { bytes =>
-            val v = rlpDecode[ByteString](bytes)
-            if (v.nonEmpty && v.length < 32) {
-              Right(MptValue(v))
-            } else {
-              Left(MptHash(v))
-            }
+          MptBranch(rlpList.items.take(BranchNodeChildLength).map {
+            case bytes: RLPValue =>
+              val v = rlpDecode[ByteString](bytes)
+              if (v.nonEmpty && v.length < 32) {
+                Right(MptValue(ByteString(rlpEncode(bytes))))
```

Comment:
yes it can be branch node so RLP list of 17 elements in which branches have to be values if node is less then 32 bytes, so it is also possible for branch node to have another branch node in on of its branches.

So to implement it correctly we have to handle 3 cases for branch nde:
* child is RLP value with hash
* child is 2 elements RLP list
* child is 17 elements RLP list

and 2 cases for extension node:
* child is RLP value with hash
* child is 17 elements RLP list